tomcat: rewrite module

aanderse opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
The current `tomcat` module suffers from a few shortcomings, such as
- it [doesn't expose configuration of server-wide components such as Engine or Connector](https://github.com/NixOS/nixpkgs/issues/61225)
- state, configuration, logging, and applications are all stored in the same stateful directory which then necessitates less than desirable options like [purifyOnStart](https://github.com/NixOS/nixpkgs/pull/44303)
- the code is extremely difficult to read and reason through properly, especially for those new to `nix` and NixOS
- how to use the module is confusing and [not well understood](https://discourse.nixos.org/t/nixops-and-tomcat/1472), deviating from NixOS norms
- the module would benefit from a section in the manual explaining some basic use cases and how to configure them, [avoiding potential confusion](https://github.com/NixOS/nixpkgs/issues/79963)
- to enable `ssl` the user is encouraged to place a secret (keystore or certificate password) on the nix store

###### Things I'm proposing
A rewrite of the module which lists `tomcat` use cases and then provides solutions to these use cases, addressing the above mentioned shortcomings of the current module.

After identifying some common use cases of `tomcat` that are relevant to me I have taken a very rough (and extremely incomplete) first pass at rewriting the module. Many options `description` values are empty or incorrect (copy+paste from another module), but I wanted to start a conversation with interested parties to see how we might improve `tomcat` on NixOS.

Here is what I came up with:
- local development, ssl doesn't matter, you want to serve over port `8080` as is the default expressed in `server.xml` from upstream
```
# quick and easy, up and running on port 8080, the developer can work from appBase
services.tomcat.enable = true;
services.tomcat.virtualHosts.localhost = {
  appBase = "/var/www/tomcat";
};
```
- run `tomcat` as a full featured web server on ports 80 and 443, PEM certificates provisioned by Let's Encrypt, requiring no keystore password in the nix store
```
# full Let's Encrypt integration out of the box, easy to host multiple sites, sane defaults but power to override most settings through various module options
services.tomcat.enable = true;
services.tomcat.virtualHosts."blog.example.org" = {
  enableACME = true;
  forceSSL = true;
  appBase = localpkgs.my-blog-war-file;
};
services.tomcat.virtualHosts."forum.example.org" = {
  enableACME = true;
  onlySSL = true;
  appBase = "/var/www/forum.example.org";
  sslHostConfig = {
      disableCompression = false;
      disableSessionTickets = true;
      honorCipherOrder = false;
  };
};
```
- serve java content through a web server like `nginx` or `apache-httpd` via `mod_jk` or reverse proxy, where `tomcat` is installed on the same machine as your web server
```
# I'm not so familiar with AJP so I need to research this more
services.tomcat.enable = true;
services.tomcat.connectors = [
  { port = 8080; protocol = "HTTP/1.1"; }
  { port = 9009; protocol="AJP/1.3"; }
];
services.tomcat.virtualHosts.localhost = {
  appBase = "/webapps";
};
# configure your web server to connect to tomcat over the appropriate port, depending on reverse proxy or mod_jk
```
- run several instances of `tomcat` over multiple servers and use something like `apache-httpd` or `nginx` to load balance them, possibly using a self signed certificate
```
# example to follow
```

###### What I want from you

@danbst @pvgoran @deliciouslytyped @tomberek @abathur as potentially interested individuals
Opinions, thoughts, comments, your use cases, module requests, examples, discussion of what I'm proposing is a good path forward or not. Minor review of code so far if you feel like it, keeping in mind it outs **not** complete. I have tested a few workflows, but not all yet.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

aanderse committed:
nixos/tomcat: rewrite module

abathur commented:
Responding so you know I'm not ignoring you :)

I don't think I've used Tomcat via Nix or otherwise, but guessing you found me via #17879. Not sure I'll have much to add directly (since my interest is just transitively in finding out if Guacamole is a viable Teamviewer alternative), but @Shados and @jraygauthier may have thoughts arising from #48140.

pvgoran commented:
The `tomcat` module is, indeed, quite limited, and arguably flawed, and rewriting it seems like a good idea. But I see multiple challenges/problems here.

1. First and foremost, existing configurations should remain working as they are, for at least one release cycle. Since the module's logic is completely replaced, probably the best way to achieve this would be to leave the existing module in place, and create a new one. (I'm not sure if this fits the establishes practices of NixOS.)
2. The old module uses tomcat's default configuration files and modifies them. The new module rebuilds the configuration files from scratch. There is a danger here that a new version of Tomcat may introduce some important contents in the default configuration file, NixOS may not notice this and use the old contents, and NixOS's tomcat instances would then lack this (presumingly important) configuration. I know that this is highly hypothetical scenario, but still it's possible, and has a potential of causing big problems.
3. Minor but important: the new module must escape XML attribute names and values. (Maybe introduce a new escaping function in `lib/strings.nix`?)
4. The location of Tomcat's stateful data shouldn't change for existing installations. (It won't be a problem if the old module is left in place, as suggested in (1).) At least, Tomcat should continue to write its logs to `/var/tomcat/log`, and store its session files in `/var/tomcat/work`.
4a. It would be good to allow specifying the location of stateful data (at least logs) via attributes. At the same time, it would to *read* these locations. (For example, I'm using something like `"${config.services.tomcat.baseDir}/logs"` to determine where the logs are, and to configure my log rotation scripts; after upgrade, I would use `config.services.tomcat.logsDir`.)
5. Minor: I don't like unconditional use of `tomcat-native`. Maybe it doesn't hurt, but I'm used to steer away from it. (It uses special kinds of connectors with unfamiliar, and possibly undesirable, properties.)
6. I'm using custom `catalinaOpts` and `commonLibs`, so I would like to see them configurable. (For completeness, `sharedLibs` and `javaOpts` should be configurable as well.)

(Also, the descriptions of `extraConfig` and `logFormat` currently refer to Apache.)

I don't have much time available, but I would like to follow the development of this module rewrite, and to help (at least by commenting/reviewing/testing).

aanderse commented:
@pvgoran thanks for taking the time to comment. A couple questions:
- are you using the `tomcat` module in production?
- are you using `tomcat` directly as a web server, or are you using it with a reverse proxy, etc...?
- are you storing any secrets in the nix store with this module?
- what connectors and ports are you using?

pvgoran commented:
> are you using the tomcat module in production?

Yes.

> are you using tomcat directly as a web server, or are you using it with a reverse proxy, etc...?

With a reverse proxy which lives on another machine connected via VPN. I'm also making local requests directly to Tomcat.

> are you storing any secrets in the nix store with this module?

I don't think so. (SSL is handled by the reverse proxy.)

> what connectors and ports are you using?

HTTP on 8080.

danbst commented on nixos/modules/services/web-servers/tomcat/default.nix:
```diff
@@ -0,0 +1,480 @@
+in
+{
+  meta = {
+    maintainers = with maintainers; [ danbst ];
```
this can be removed, I no longer maintain Tomcat installations.

aanderse replied:
OK.

danbst commented on nixos/modules/services/web-servers/tomcat/default.nix:
```diff
@@ -0,0 +1,480 @@
+  # TODO: write a custom logging.properties file i guess
+  # https://github.com/GoogleCloudPlatform/tomcat-runtime/pull/28/files
+  loggingProperties = pkgs.writeText "logging.properties" ''
+    handlers = java.util.logging.ConsoleHandler
```
yes, would be nice for this to be configurable. I had this replaced by
```
             # this configures such tomcat log files:
             # - catalina.log: contains all tomcat messages
             # - catalina.out: all System.out and System.err messages
             # catalina.log and catalina.out have no rotation, so logrotate must be used to limit files by size
             # (rumors on internet say tomcat crashes when log files are > 2G)

             handlers = 1catalina.org.apache.juli.AsyncFileHandler
             .handlers = 1catalina.org.apache.juli.AsyncFileHandler

             1catalina.org.apache.juli.AsyncFileHandler.level = FINE
             1catalina.org.apache.juli.AsyncFileHandler.directory = $\{catalina.base}/logs
             1catalina.org.apache.juli.AsyncFileHandler.rotatable = false
             1catalina.org.apache.juli.AsyncFileHandler.prefix = catalina
```

aanderse replied:
Now fully configurable, but defaults to logging everything except access logs to `journald` :+1:

danbst commented on nixos/modules/services/web-servers/tomcat/default.nix:
```diff
@@ -0,0 +1,480 @@
+
+    services.tomcat.extraConfig = ''
+      <Listener className="org.apache.catalina.startup.VersionLoggerListener" />
+      <Listener className="org.apache.catalina.core.AprLifecycleListener" SSLEngine="on" />
```
actually, I had this specific line removed in my config:
```
              # disable tomcat native library (we don't want use tomcat as web server)
              sed -i '/<Listener className="org.apache.catalina.core.AprLifecycleListener" SSLEngine="on" \/>/d' $out/conf/server.xml
```
AFAIK the only reason to do this was to remove some warnings from logs.

aanderse replied:
I'm going to look into this a bit more, but for starters I have made a `listeners` option of type `list` to declore `<Listener>` elements, allowing for much finer control if desired. As you mention a simple fix would be to conditionally add this depending on whether `tomcat-native` has been included.

danbst commented on nixos/modules/services/web-servers/tomcat/default.nix:
```diff
@@ -0,0 +1,480 @@
+          ${robotsEntries}
+          <Valve
+            className="org.apache.catalina.valves.AccessLogValve"
+            directory="${logsDir}"
```
in my case I had this commented out, because tomcats was behind nginx, so we had already access logs:
```
              # disable access log (we use nginx instead)
              sed -i -e 's/<Valve className="org.apache.catalina.valves.AccessLogValve"/<!--Valve className="org.apache.catalina.valves.AccessLogValve"/g' \
                     -e '/;%r&quot; %s %b" \/>/a -->' \
                  $out/conf/server.xml
```

aanderse replied:
Makes sense. I made `virtualHosts.logFormat` of `type` `nullOr str` so that when `logFormat = null;` is passed no logging will occur. Thanks for the suggestion :+1:

aanderse commented:
Thanks for your comments @danbst. I've had some higher priority stuff get in the way of this lately but will eventually circle back around.

aanderse commented:
@pvgoran thanks for your feedback. I've finally come back around to this PR. While still heavily WIP I'm making some progress and can somewhat address your points now.

1. I understand. Fortunately NixOS makes it trivial to pin modules so you can take as long as you need to migrate. See https://nixos.org/nixos/manual/#sec-replace-modules for details if you aren't already familiar with this.
2. `tomcat` is established and stable enough that dramatic changes aren't frequent. `diff` is a pretty solid tool for maintainers to guard against the type of problems you mention. Building `server.xml` the way this new module does is much more flexible and just as robust as the use of `sed` (which is scary! :scream:) in the current module.
3. That could be useful, though not strictly required. Something to keep in mind.
4. I'm glad you mention the logs here. I hadn't added this option into my rewrite yet, but definitely should have. I have now added a `logDir` option which will store `tomcat` access logs. I'm still looking into the implications of the `work` directory. The way the current module copies configuration into stateful directories is **bad**. The new module has separated configuration from stateful data so configuration can be truly immutable, which is not only much cleaner, but a big win.

5. You are definitely correct - this shouldn't be forced upon users who don't want/need it. I have fixed this now. The idea is that if you are using `tomcat` as an actual web server (as opposed to pairing with `nginx`/`httpd`) you want to pull in `tomcat-native`, otherwise you don't.

6. `catalinaOpts` and `javaOpts` have been added back in :+1:. Upstream suggests **NOT** using `sharedLibs` and `commonLibs` to a point where they have removed the way this worked in `tomcat` versions `6` or `7`. Upstream suggests including `jar` files in your web application, which seems to fit the nix way of doing things reproducibly (build your apps with `maven`/`gradle` and specify all dependencies exactly). There is still the ability to do this (and I'll document it, it is relatively elegant) using `catalina.properties` but I don't think as a distro we want to promote things upstream is suggesting users steer clear of.

More to come.

pvgoran commented:
@aanderse I didn't find time to reply yet. :(
But I'm going to do it.

Mic92 commented on nixos/modules/services/web-servers/tomcat/default.nix:
```diff
@@ -0,0 +1,554 @@
+      ''
+  ;
+
+  serverXmlFile = pkgs.writeText "server.xml" ''
```
by the way `toXML` does not work for cases like this?

aanderse replied:
I'll have to give that a try. Thanks @Mic92!

Mic92 commented:
There is an temporary editor file committed that should be removed as well.

aanderse commented:
Unfortunately I don't have reason to finish this.
